Some parameters will be overwritten which was configured in consumer.config.

SweetMojito opened the pull request:
Some parameters will be overwritten which was configured in consumer.config where running "ConsumerPerformance.scala"
linked see https://issues.apache.org/jira/browse/KAFKA-10043

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [x] Verify design and implementation 
- [x] Verify test coverage and CI build status
- [x] Verify documentation (including upgrade notes)

SweetMojito committed:
Update ConsumerPerformance.scala

Some parameters will be overwritten which was configured in consumer.config where running "ConsumerPerformance.scala"
linked see https://issues.apache.org/jira/browse/KAFKA-10043

itantiger commented:
> Could you update the code (according to the reviews) and then push the changes to be a new commit. Also, please rebase your code (by trunk).
@chia7712  Can you take a look at this?

SweetMojito committed:
optimized code

chia7712 commented:
Could you add test for this issue?

itantiger commented:
> Could you add test for this issue?

Do you mean to add unit tests？But i did not found unit tests for this class `kafka.tools.ConsumerPerformance.scala`

chia7712 commented:
> But i did not found unit tests for this class kafka.tools.ConsumerPerformance.scala

It would be great if you could be the first person to write the tests for  ```ConsumerPerformance``` :)

SweetMojito committed:
add unit test of #9790

add unit test of #9790

itantiger commented:
> > But i did not found unit tests for this class kafka.tools.ConsumerPerformance.scala
> 
> It would be great if you could be the first person to write the tests for `ConsumerPerformance` :)

I had committed the unit test in `ConsumerPerformanceTest`, What should I do to relate to that?

itantiger commented:
> > > But i did not found unit tests for this class kafka.tools.ConsumerPerformance.scala
> > 
> > 
> > It would be great if you could be the first person to write the tests for `ConsumerPerformance` :)
> 
> I had committed the unit test in `ConsumerPerformanceTest`, What should I do to relate to that?

I found that it was automatically linked，but all checks had failed :( .
what should I do next :)

chia7712 commented:
> what should I do next :)

please take a look at the report and fix the error. https://github.com/apache/kafka/pull/9790/checks?check_run_id=1616222522

SweetMojito committed:
add import

itantiger commented:
> > what should I do next :)
> 
> please take a look at the report and fix the error. https://github.com/apache/kafka/pull/9790/checks?check_run_id=1616222522

I fixed the error, but there were many errors in the second build. It seemed that the errors had nothing to do with what I changed. Please help me to have a look :)

chia7712 commented:
@itantiger Thanks for your patch. Could you revise the title?

itantiger commented:
> @itantiger Thanks for your patch. Could you revise the title?

Also thank you very much for your guidance :) .

itantiger commented:
> @itantiger Thanks for your patch. Could you revise the title?

Finished.
When will this patch merge to trunk?

chia7712 commented on core/src/test/scala/unit/kafka/tools/ConsumerPerformanceTest.scala:
```diff
@@ -96,6 +97,28 @@ class ConsumerPerformanceTest {
+    val propsStream = Files.newOutputStream(propsFile.toPath)
+    propsStream.write("group.id=test_group_id\n".getBytes())
+    propsStream.write("client.id=test_client_id".getBytes())
+    propsStream.close()
```
Could you test ```--topic``` to make sure it is not overrided by command-line arguments.

SweetMojito replied:
`--topic` is a `REQUIRED` argument in command-line, I thank Its value is determined by the command line.

SweetMojito replied:
So, if config file has the param `topic=test`, It will not take effect.

chia7712 replied:
Or could we add ```group.id``` to command-line arguments to make sure it (from file) is NOT override by command-line args?

SweetMojito replied:
Okey, I finished the local test and submitted the code :)

SweetMojito replied:
@chia7712  Please take a look :) Is there anything else that needs to be optimized？

chia7712 reviewed (commented):
@itantiger Thanks for your patch. one small comment is left. Please take a look :)

SweetMojito committed:
test of add group.id to command-line arguments

add group.id to command-line arguments to make sure it (from file) is not override by command-line args.

chia7712 reviewed (approved):
@itantiger Thanks for your great patch. LGTM. Will merge it tomorrow :)

chia7712 commented:
@itantiger The priority of loading props is a bit chaos in this case. It seems to me the property of loading properties is shown below.
1. command-lind
1. config file (consumer.config)
1. default value

The current priority is shown below.
1. command-lind
1. default value
1. config file (consumer.config)

The priority from this PR is shown below.
1. config file (consumer.config)
1. command-lind
1. default value


Of course, current priority has issue as properties from config file always get overwritten by default value. We need to fix it. However, I'd like to check the priority again. WDYT? Should config file has higher priority than command-lind?

itantiger commented:
> 1. command-lind
> 2. config file (consumer.config)
> 3. default value
I think this priority is more better, and this PR is to do this.

chia7712 commented:
> I think this priority is more better, and this PR is to do this.

We load the props from config file first. https://github.com/apache/kafka/blob/afa5423356d3d2a2135a51200573b45d097f6d60/core/src/main/scala/kafka/tools/ConsumerPerformance.scala#L275 and then call ```props.putIfAbsent``` to set args from command-line only if the prop is nonexistent in config file. Hence, the priority of config file is higher than command-line, right?

itantiger commented:
> and then call `props.putIfAbsent` to set args from command-line only if the prop is nonexistent in config file. Hence, the priority of config file is higher than command-line, right?
You're right. I've ignored it. It should be modified here. :)

itantiger commented:
> > and then call `props.putIfAbsent` to set args from command-line only if the prop is nonexistent in config file. Hence, the priority of config file is higher than command-line, right?
> > You're right. I've ignored it. It should be modified here. :)

itantiger commented:
Current priority, some config params would be overwritten by default value, Do you think it needs to be revised?

chia7712 commented:
> Do you think it needs to be revised?

I guess you are talking about "title". yes, the title need to be revised.

itantiger commented:
> I guess you are talking about "title". yes, the title need to be revised.
In addition, I think the trunk also needs to be modified.
Some config params would be overwritten by default value, WDYT?

itantiger commented:
> I guess you are talking about "title". yes, the title need to be revised.

In addition, I think the trunk also needs to be modified.
Some config params would be overwritten by default value, WDYT?

chia7712 commented:
> I think the trunk also needs to be modified.

trunk?  pardon me that I can't catch your point.

> Some config params would be overwritten by default value, WDYT?

yep, the default value CAN overwrite the configs from either command-line or config file.

itantiger commented:
> >trunk? pardon me that I can't catch your point.
> That's my problem. hah
>> yep, the default value CAN overwrite the configs from either command-line or config file.
> Let me optimize it again and commit. Do you think it's necessary?

itantiger commented:
>>trunk? pardon me that I can't catch your point.
>
>That's my problem. hah
>
>>yep, the default value CAN overwrite the configs from either command-line or config file.
>
>Let me optimize it again and commit. Do you think it's necessary?

chia7712 commented:
> Let me optimize it again and commit. Do you think it's necessary?

I'd like to address the following priority rule in this PR :)

1. command-lind
2. config file (consumer.config)
3. default value

itantiger commented:
> I'd like to address the following priority rule in this PR :)
> 
> 1. command-lind
> 2. config file (consumer.config)
> 3. default value

Of course, I also think this is more reasonable
